### PR TITLE
[WIP] [OptionsResolver] handle nested options

### LINKED ISSUE #4833: [OptionsResolver] Support nested options
It is currently not possible to configure nested options. For example:

``` php
$resolver->setDefaults(array(
    'db' => array(
        'dsn' => null,
        'user' => 'root,
        'password' => '',
        'port': 3306,
    ),
));
```

It is not possible to resolve this configuration in the following way:

``` php
$resolver->resolve(array(
    'db' => array(
        'dsn' => ...
    ),
));
```

because now the other default values under "db" will be dropped. Also, it is not possible to specify allowed values or types for the nested options.

### LINKED ISSUE #15524: [OptionsResolver] array of instance allowed type
It could be nice to have possibility to make this kind of requirement:

``` php
->setAllowedTypes('attachments', '\Swift_Attachment[]')
```

Basically, for this case, an array of `Swift_Attachment` objects.

It could be a good beginning waiting #4833.

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #4833, #15524 |
| License | MIT |
| Doc PR | todo |
- [ ] Gather feedback
- [ ] Address a doc PR
- [ ] Update CHANGELOG
- [ ] Add a lot of tests :)
# Summary
- c2944ac fixes #4833. Handle nested options resolving
  - add `OptionsResolver::nested` a private array of nested `OptionsResolver` instances by root names,
  - add `OptionsResolver::setNested($option, array $defaults)` defines an option as a new self and returns the nested `OptionsResolver` instance,
  - add `OptionsResolver::isNested($option)` returns whether an option name is a root name for a nested `OptionsResolver` instance,
  - add `OptionsResolver::getNested()` an array of root names representing nested `OptionsResolver` instances.
  
  ``` php
  $connexionResolver = $resolver->setNested('connexion', array(
      'host' => 'localhost',
      'port' => 80,
  );
  
  $connexionResolver->setRequired('user');
  $connexionResolver->setDefault('password', function (Options $connexion) {
      return isset($nested['user']) ? '' : null;
  });
  $connexionResolver->setDefault('secure', false);
  $connexionResolver->setNormalizer('secure', function (Options $connexion, $secure) {
      return 443 === $nested['port'] ?: $secure;
  });
  $connexionResolver->setAllowedTypes('port', 'int');
  $connexionResolver->setAllowedTypes('user', 'string');
  
  $resolver->setRequired('connexion');
  $resolver->setNormalizer('connexion', function (Options $options, $resolvedConnexion) {
      if ('localhost' === $resolvedConnexion['host']) {
          $resolvedConnexion['host'] = $option['param'];
      }
  
      return $resolvedConnexion;
  });
  
  // In a sub class
  parent::configureOptions($resolver);
  
  $resolver->setDefault('user', $user);
  $connexionResolver = $resolver->setDefault(
      'connexion', function (Options $options, $previousDefaultConnexion) {
          if ($options['user'] instanceof UserInterface and 'localhost' !== $previousDefaultConnexion['host'])
              $connexion['user'] = $options['user']->getUsername();
          }
  );
  
  $options = $resolver->resolve($array);
  ```
  
  `*Allowed*` methods are not supported with root name of nested options since they always have to be an array, so:
- f2ad947 trigger warning ? throw `BadMethodCallException` ? what do you think ?
- 9e4e02a fixes #15524. Handle extra options resolving and `addAllowed*ForAll` rules with constant flags
  - add `Options::NONE` constant,
  - add `Options::ALL` constant, matches defined and extra options,
  - add `Options::DEFINED` constant,
  - add `Options::NESTED` constant,
  - add `Options::EXTRA` constant, matches any undefined options, helps dealing with "anonymous" options (using a numerical index)
  - add `OptionsResolver::allowedValuesForAll` a private array to validate all options even undefined but not root of nested options as they must always be an array,
  - add `OptionsResolver::allowedTypesForAll` a private array to validate all options (as previous),
  - add `OptionsResolver::resolveUndefined` a private `bool` to allow extra options resolving,
  - add `OptionsResolver::addAllowedValuesForAll($options, $allowedValues, $replace = false, $nested = Options::ALL)`, first argument may be an array of options names or any flag constant except `Options::NONE`, if a root name of nested options is passed as/among first argument the method will be recursive because the fourth argument default is `Options::ALL`, prevent it by passing `Options::NONE` instead,
  - add `OptionsResolver::addAllowedTypesForAll($options, $allowedTypes, $replace, $nested)`, same as previous, note that passing `Options::ALL` as first argument will not apply to root of nested options as they must always be arrays and on contrary will apply on extra options, pass `true` as third argument to override previous allowed types,
  - add `OptionsResolver::setPrototypes((string|string[]) $types, (mixed) $values, $options = Options::EXTRA)`, sets `resolveUndefined` to `true` and validate any extra options with the any given type with its corresponding allowed values,
  - add `OptionsResolver::allowExtraOptions($allow = true)`
  
  ``` php
  $resolver->setNested('attachments', array($attachments))
      ->allowExtraOptions()
      ->addAllowedTypesForAll(Options::ALL, '\Swift_Attachment');
  
  // Another way by implicitly allowing extra options:
  $resolver->setNested('emails')
      ->setPrototypes('string', function ($email) use ($regex) {
          return preg_match($regex, $email);
      });
  ```
  
  The difference is that in the second case implicitly the flag `Options::EXTRA` is passed so previous defaults will not be validate by prototypes rules, if that's not what you want pass `Options::ALL` as third argument.
  Note that `setNested()` returns the nested `OptionsResolver` instance in this examples, you could call these methods on the `$resolver` as well but this not the same:
  
  ``` php
  // Any undefined global options can only be strings
  $resolver
      ->allowExtraOptions()
      ->addAllowTypesForAll(Options::EXTRA, 'string', true, Options::NONE);
  ```
- ccd5ce1 Harden dependency of nested options with an internal class
  - add `NestedOptions` class which extends `OptionsResolver`,
  - add `NestedOptions::root` the parent `OptionsResolver` instance,
  - add `NestedOptions::rootName` the root option name,
  - add `NestedOptions::resolve()` overrides to throw an exception is parent `OptionsResolver` instance is not locked,
  - add `OptionsResolver::isLocked()` internal public method to check from `NestedOptions` if it can be resolved.
# Work in Progress (RFC)
- add `OptionsResolver::setExtraNormalizer(function (Options $options, $extraName, $extraValue) { ... })` to normalize any undefined options, basically a shortcut for the currently supported `setNormalizer(Options::Extra, callable)` but with handling of `$extraName`
- add a `OptionsResolver::prototypes` private array holding allowed values by allowed types to validate extra options.
- be more strict on NestedOptions array type and indexing when setting defaults. Currently the implementation does not distinguish  nested defaults using a numerical index instead of names, which may cause problem when overriding them or to properly map validation/normalization.

### CODE REVIEW

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 1]
```diff
@@ -488,6 +600,12 @@ public function addAllowedValues($option, $allowedValues)
             throw new AccessException('Allowed values cannot be added from a lazy option or normalizer.');
         }
 
+        if ($this->isNested($option)) {
+            @trigger_error(sprintf('The "%s" method should not be used with nested options. A failed attempt occurred with the option "%" for values %s', __METHOD__, $option, $this->formatValues($allowedValues)), E_USER_WARNING);
```
Comment (Ma27) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: this doesn't seem to be a deprecation notice, so exceptions should be used, right?

Comment (HeahDude) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: That what I proposed in my [comment](https://github.com/HeahDude/symfony/commit/49588c6a053f717f5e14b383e235437a106c9bf8#commitcomment-16664140) in the concerned commit.

I'm not sure how strict we should be here, this is just a proposition.

**Comment by HeahDude** [comment]
I've made some little refactoring and updated the description, ready for another round of reviews.

Still need to add a lot of tests.

Cheers

**Comment by HeahDude** [comment]
I tried to click on the heart by @sstok but that added one, this one does not count! (but is true :)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 2]
```diff
@@ -593,6 +740,71 @@ public function addAllowedTypes($option, $allowedTypes)
...
+     *                                          for nested options which is the
+     *                                          default. Pass Options::NONE to
+     *                                          change it.
+     *
+     * @return OptionsResolver This instance
+     *
+     * @throws UndefinedOptionsException If an option is undefined
+     * @throws AccessException           If called from a lazy option or normalizer
+     */
+    public function addAllowedTypesForAll($optionNames, $allowedTypes, $replace = false, $nested = Options::ALL)
```
Comment (HeahDude) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Should I do the same with `allowedValuesForAll()` ? There might be some use cases.

Comment (HeahDude) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Done in d54ca59.

**Comment by HeahDude** [comment]
Pushed 211a4d9 for handle extra options validation separately from the others.

Description updated.

**Comment by HeahDude** [comment]
I've squashed things up and pushed ccd5ce1 for an attempt to wrap nested options is a safe and internal way.

Updated the description to be more accurate.

**Comment by HeahDude** [comment]
Pushed  e8f1172 to pass the parent options to each nested option normalizer.

**Comment by backbone87** [comment]
Just an alternative approach for this feature, without caring about maintaining a tree:
- Accept `PropertyPath`s as option names
- Provide a wrapper for `OptionsResolver` that automatically converts option names to `PropertyPath`s

```
$resolver->setDefault(new PropertyPath('connexion[secure]'), false);
$wrapped = new PropertyPathOptionsResolverDecorator($resolver);
$wrapped->setDefault('connexion[secure]', false); // same as 2 lines above
```

**Comment by backbone87** [comment]
Just some more thoughts: Your approach of maintaining a tree makes the `OptionsResolver` almost a clone of `Configuration`

**Comment by HeahDude** [comment]
@backbone87 thanks for the feedback.

>Accept PropertyPaths as option names
Provide a wrapper for OptionsResolver that automatically converts option names to PropertyPaths

It's not a bad idea. But for now I'd rather try to handle nested or extra options without adding too much overhead to the current behavior.

>Just some more thoughts: Your approach of maintaining a tree makes the OptionsResolver almost a clone of Configuration

I don't agree with you on that, it just makes it more powerful IMHO, ref #3968.

I've worked on that PR but haven't push any more updates for now, I'm waiting for more reviews on the current state.

ping @stof @Tobion @webmozart

**Comment by backbone87** [comment]
> I don't agree with you on that, it just makes it more powerful IMHO, ref #3968.

Well that discussion was 4 years ago and the OptionsResolver grew quite a bit since then **and** the requested features push it more and more towards the feature set of the Config component

**Comment by HeahDude** [comment]
But they still do not share goal or responsibility: one for loading resources and defining merging rules for domain configuration and the other for resolving class options even through inheritance.

**Comment by backbone87** [comment]
"*option resolution through inheritance*" is a specific "*defined merging rule*"
the "*configuration domain*" of the Form component are `FormType`s

The main benefit of the `OptionResolver` is probably much better performance, but thats something that can be worked out in the Config component.

The point of view is cleary personal preference, so like you already said: lets hear some other opinions on this topic.

**Comment by HeahDude** [comment]
I would just say I see a nuance between "domain configuration" and "component or classes configuration" which may not require external resource loading but internal definition.

**Comment by webmozart** [comment]
Thank you for working on this @HeahDude! I'm a bit uncomfortable about adding this functionality. First of all, nested options - so far - were not asked for by many people. Second, adding support for nested options *will* have a performance impact. However - just like PropertyAccess - OptionsResolver is a very low-level but frequently called component that impacts the performance of forms. We should avoid decreasing its performance for no benefit (for the majority of users).

**Comment by HeahDude** [comment]
@webmozart thanks for the review. It was my self challenge for the hack day.

It looks more like a PoC/RFC as is, and would require a nice lifting and lot of test to be safely merged right now.

I actually thought about the light weight aspect, but the idea is also to help some core form types.

Since it remains a specific new method call, it should not impact performance that much while being flexible for specific use cases, `attr` is a perfect example.

Also it may be more safe to allow "locked" options as tried in #17161 since some options should never been overridden in some form type extensions or children in some core form types.

I guess we have the time to polish it for later and make a decision when it's properly implemented.

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 3]
```diff
@@ -593,6 +839,159 @@ public function addAllowedTypes($option, $allowedTypes)
...
+     * @param int                 $nested       This method is recursive for
+     *                                          nested options which is the
+     *                                          default. Pass Options::NONE
+     *                                          to change it.
+     *
+     * @return OptionsResolver This instance
+     *
+     * @throws UndefinedOptionsException If an option is undefined
+     * @throws AccessException           If called from a lazy option or normalizer
+     */
```
Comment (trousers) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Codacy Issue found: [The method addAllowedTypesForAll() has a Cyclomatic Complexity of 12. The configured cyclomatic complexity threshold is 10.](https://www.codacy.com/app/git_9/symfony/file/2332052707/issues/source?bid=3207821&fileBranchId=3207869#l875)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php [hunk 4]
```diff
@@ -513,6 +655,98 @@ public function addAllowedValues($option, $allowedValues)
...
+     * @param int                 $nested        This method is recursive for
+     *                                           nested options which is the
+     *                                           default. Pass Options::NONE
+     *                                           to change it.
+     *
+     * @return OptionsResolver This instance
+     *
+     * @throws UndefinedOptionsException If an option is undefined
+     * @throws AccessException           If called from a lazy option or normalizer
+     */
```
Comment (trousers) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: Codacy Issue found: [The method addAllowedValuesForAll() has a Cyclomatic Complexity of 12. The configured cyclomatic complexity threshold is 10.](https://www.codacy.com/app/git_9/symfony/file/2332052707/issues/source?bid=3207821&fileBranchId=3207869#l699)

File: src/Symfony/Component/OptionsResolver/OptionsResolver.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Component/OptionsResolver/OptionsResolver.php]: @trousers Please stop blindly posting analysis results in every pending pull request.

**Comment by HeahDude** [comment]
Status: Needs Work

**Comment by senkal** [comment]
We have an use case where we need this functionality.
For now, we just created some class later which provides logic we need,  instead of using OptionResolver directly.
Maybe an option would be to move it to higher level(the decision if we want to use the potentially slower   option nested resolver), than to change main OptionResolver code.
Maybe new thing, something like collection of those resolver, something similar to ValidatorConstraint/All http://symfony.com/doc/current/reference/constraints/All.html?
Looks like, from what I have read, this is common request to handle just one more level.
Anyway, that wouldn't support anything more than one more level.
In the end, looks like, for majority not needed.
For very deeply nested arrays- writing your own solution is not that bad.

**Comment by fabpot** [comment]
@HeahDude Any news on this one? Still wanting to work on it?

**Comment by HeahDude** [comment]
Yes I think it's needed to support this case IMHO even in the core. I will start it again from scratch to get something simpler and be careful to not add too much overhead in the meantime if someone wants to give it a try, please do.

Closing for now.

**Pull request closed without merging**